File: TraningAPI/RealmDB/RealmDB.m
//
//  RealmDB.m
//  TraningAPI
//
//  Created by 高橋慧 on 2021/06/18.
//  Copyright © 2021 k-takahashi. All rights reserved.
//

#import "PrefecturesEntity.h"
#import "RealmDB.h"

@implementation RealmDB
static RealmDB *shared_ = nil;

+ (RealmDB *)sharedManager {
   @synchronized(self){
       if (!shared_) {
           shared_ = [RealmDB new];
       }
   }
   return shared_;
}

- (id)init
{
   self = [super init];
   if (self) {
       self.realm = [RLMRealm defaultRealm];
   }
   return self;
}
- (void)setPrefecturesOnRealmDB:(NSArray<Prefectures *> *)prefectures {
    [_realm beginWriteTransaction];
    [_realm addObjects: prefectures];
    [_realm commitWriteTransaction];
}
- (void)updatePrefecturesOnRealmDB:(NSArray<Prefectures *> *)prefectures {
    [_realm beginWriteTransaction];
    [_realm addOrUpdateObjects: prefectures];
    [_realm commitWriteTransaction];
}
- (NSArray<Prefectures *> *)getPrefecturesByRealmDB {
    RLMResults<Prefectures *> *result = [Prefectures allObjects];
    NSMutableArray *tempPrefectures = [NSMutableArray array];
    for (Prefectures *data in result) {
        [tempPrefectures addObject: data];
    }
    NSArray<Prefectures *> *prefectures = tempPrefectures;
    return prefectures;
}
- (Prefectures *)getPrefecturesByID: (int)idNubmer {
    id idObject = [NSNumber numberWithInt:idNubmer];
    Prefectures *result = [Prefectures objectInRealm: self.realm forPrimaryKey:idObject];
    return result;
}
@end

File: TraningAPI/RealmDB/RealmDB.swift
//
//  RealmDB.swift
//  TraningAPI
//
//  Created by 高橋慧 on 2021/04/21.
//  Copyright © 2021 k-takahashi. All rights reserved.
//

import Foundation
import RealmSwift

final class RealmDB{
    
    //本当はDoでエラーキャッチしなくてはいけないはず
    private var realm:Realm?
    init() {
        do {
            realm = try Realm()
        }catch{
            realm = nil
        }
        
    }
    
    //    MARK: - Create Record　レコードを追加
    ///
    /// - Parameters:
    ///   - prefectures:Prefecturesクラスの配列
    func setPrefecturesOnRealmDB(prefectures:[Prefectures]){
        do {
            try self.realm?.write {
                self.realm?.add(prefectures)
            }
        } catch {
        }
    }
    //    MARK: - Update Record　レコードを更新
    ///
    /// - Parameters:
    ///   - prefectures:Prefecturesクラスの配列
    func updatePrefecturesOnRealmDB(prefectures:[Prefectures]){
        do {
            try self.realm?.write {
                self.realm?.add(prefectures, update: .modified)
            }
        } catch {
        }
    }
    //    MARK: - Get All Record　全レコードを取得
    ///
    /// - Returns:Prefecturesクラスの配列
    func getPrefecturesByRealmDB() -> [Prefectures]?{
        var prefectures:[Prefectures]? = [Prefectures]()
        //全権取得
        if let results = self.realm?.objects(Prefectures.self){
            if results.isEmpty {
                return nil
            }else{
                results.forEach{ prefectures?.append($0)}
            }
        }
        return prefectures
    }
    //    MARK: - Get A Record　該当レコードを取得
    ///
    /// - Parameters:
    ///   - id:Prefecturesテーブルのプライマリキー
    /// - Returns:Prefecturesクラスの配列
    func getprefecturesByID(id: Int) -> Prefectures?{
        //プライマリキーで取得
        if let result = self.realm?.object(ofType: Prefectures.self, forPrimaryKey: id){
            return result
        }
        return nil
    }
}
